README.md
# Practice

# ImageCropper

A simple app that does the following:

1) A web panel where the user can upload an image
2) Each image has a recommended size of 1024 x 1024.
3) Don’t upload the file if it’s not the right size.
4) Each image has to be converted into 4 different sizes. 
horizontal : 755 x 450
vertical : 365 x 450
horizontal small : 365 x 212
gallery : 380 x 380
Images should not be stretched, and they should be cropped properly. 

5) Save all 4 of these images locally on the server
6) Show a webpage with all 4 of these new images.

Bonus : Instead of saving these files to the server, upload them to a
cloud image hosting service.

Double Bonus : While uploading the image show a preview in the
browser itself of all the different image sizes, and let the user
decide how to crop the images to the smaller size.

---------------------------------------------------------------------------

# IP Address Finder

Examine the Sample Input below:

input = ["1233454356\tGET /index.html\t10.10.10.1\n",
"1233454356\tGET /index.html\t10.10.10.1\n",
"1233454356\tGET /index.html\t10.10.10.1\n",
"1233454356\tGET /index.html\t10.13.10.1\n",
"1233454356\tGET /index.html\t10.14.10.1\n",
"1233454356\tGET /index.html\t10.14.10.1\n",
"1233454356\tGET /index.html\t10.14.10.1\n",
"1233454356\tGET /index.html\t123.14.10.1\n",
"1233454356\tGET /index.html\t10.14.10.1\n",
"1233454356\tGET /index.html\t10.14.10.1\n",
"1233454356\tGET /index.html\t10.14.10.1\n",
"1233454356\tGET /index.html\t10.14.11.1\n",
"1233454356\tGET /index.html\t10.14.11.1\n",
"1233454356\tGET /index.html\t10.14.11.1\n",
"1233454356\tGET /index.html\t10.14.11.1\n",
"1233454356\tGET /index.html\t10.14.11.1\n",
"1233454356\tGET /index.html\t10.14.10.1\n",
"1233454356\tGET /index.html\t123.14.10.1\n"]

Your task is to write javascript code to find the top ten IP address and print them in following format:

Sample output:

RANK    Input               COUNT
1       10.14.10.1          7
2       10.14.11.1          5
3       10.10.10.1          3
4       123.14.10.1         2
5       10.13.10.1          1

---------------------------------------------------------------------------

#Random Quote Generator

Build an app that is functionally similar to this: https://codepen.io/freeCodeCamp/full/qRZeGZ.

---------------------------------------------------------------------------

# Auto Generated

This project was generated with [Angular CLI](https://github.com/angular/angular-cli) version 9.0.7.

## Development server

Run `ng serve` for a dev server. Navigate to `http://localhost:4200/`. The app will automatically reload if you change any of the source files.

## Code scaffolding

Run `ng generate component component-name` to generate a new component. You can also use `ng generate directive|pipe|service|class|guard|interface|enum|module`.

## Build

Run `ng build` to build the project. The build artifacts will be stored in the `dist/` directory. Use the `--prod` flag for a production build.

## Running unit tests

Run `ng test` to execute the unit tests via [Karma](https://karma-runner.github.io).

## Running end-to-end tests

Run `ng e2e` to execute the end-to-end tests via [Protractor](http://www.protractortest.org/).

## Further help

To get more help on the Angular CLI use `ng help` or go check out the [Angular CLI README](https://github.com/angular/angular-cli/blob/master/README.md).


Changelog.md
#Changes

##0.4.0
Added component & route for Markdown Previewer.
Added basic HTML & CSS for markdown Previewer.

##0.3.0
Added Random Quote Generator.
Added relevant HTML, CSS and code.
Added font awesome for Facebook & Twitter icons.

##0.2.1
Added code to extract IP Addresses from the input JSON Array.
Storing the IP Address and its respective counter in an Object.
Displaying IP Address and count in a material table.
Added index to output table.

##0.2.0
Added IP Address finder component.
Added sample JSON input file.
Added config settings in tsconfig to enable reason JSON files.
Added IP Address Finder details in Readme.

##0.1.2
Added description on home page.
Closing side menu on button click.

##0.1.1
Added code to anonymously upload a cropped image to IMGUR

##0.1.0
Added Image Cropper code.

##0.0.3
Added Angular Material.
Added Side Navigation Menu.
Modified Side Menu to be dynamic.

##0.0.2
Added route to home page.
Modified first page to be Home.

##0.0.1
Added app routing module
Added routes to router module
Added router-outlet.

src/app/quote-generator.service.spec.ts
import { TestBed } from '@angular/core/testing';

import { QuoteGeneratorService } from './quote-generator.service';

describe('QuoteGeneratorService', () => {
  let service: QuoteGeneratorService;

  beforeEach(() => {
    TestBed.configureTestingModule({});
    service = TestBed.inject(QuoteGeneratorService);
  });

  it('should be created', () => {
    expect(service).toBeTruthy();
  });
});


src/app/ip-address-finder/ip-address-finder.component.spec.ts
import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { IpAddressFinderComponent } from './ip-address-finder.component';

describe('IpAddressFinderComponent', () => {
  let component: IpAddressFinderComponent;
  let fixture: ComponentFixture<IpAddressFinderComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ IpAddressFinderComponent ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(IpAddressFinderComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
});


src/app/app.module.ts
import { NgModule } from '@angular/core';
import { HttpClientModule } from '@angular/common/http';
import { BrowserModule } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';

import { faTwitter, faFacebook } from '@fortawesome/free-brands-svg-icons';
import { FontAwesomeModule, FaIconLibrary } from '@fortawesome/angular-fontawesome';
import { MatTooltipModule, MatCardModule, MatTableModule, MatToolbarModule, MatIconModule, MatSidenavModule, MatListModule, MatButtonModule, MatExpansionModule } from  '@angular/material';

import { AppComponent } from './app.component';
import { HomeComponent } from './home/home.component';
import { AppRoutingModule } from './app-routing.module';
import { ImageCropperComponent } from './image-cropper/image-cropper.component';
import { QuoteGeneratorComponent } from './quote-generator/quote-generator.component';
import { IpAddressFinderComponent } from './ip-address-finder/ip-address-finder.component';
import { MarkdownPreviewerComponent } from './markdown-previewer/markdown-previewer.component';

@NgModule({
  declarations: [
    AppComponent,
    HomeComponent,
    ImageCropperComponent,
    IpAddressFinderComponent,
    QuoteGeneratorComponent,
    MarkdownPreviewerComponent
  ],
  imports: [
    BrowserModule,
    AppRoutingModule,
    BrowserAnimationsModule,
    MatToolbarModule,
    MatSidenavModule,
    MatListModule,
    MatButtonModule,
    MatIconModule,
    HttpClientModule,
    MatExpansionModule,
    MatTableModule,
    MatCardModule,
    FontAwesomeModule,
    MatTooltipModule
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule {
    constructor(library: FaIconLibrary) {
    // Added icons to the library for convenient access in other components
    library.addIcons(faFacebook);
    library.addIcons(faTwitter);
  }
}


src/app/app.component.ts
import { MediaMatcher } from '@angular/cdk/layout';
import { MatSidenav } from '@angular/material/sidenav';
import { ChangeDetectorRef, Component, OnDestroy, ViewChild } from '@angular/core';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements OnDestroy {
  
	mobileQuery: MediaQueryList;
	private _mobileQueryListener: () => void;
	@ViewChild('sidenav', {static: false}) sidenav: MatSidenav;

	constructor(changeDetectorRef: ChangeDetectorRef, media: MediaMatcher) {
		this.mobileQuery = media.matchMedia('(max-width: 600px)');
		this._mobileQueryListener = () => changeDetectorRef.detectChanges();
		this.mobileQuery.addListener(this._mobileQueryListener);
	}

	ngOnDestroy(): void {
		this.mobileQuery.removeListener(this._mobileQueryListener);
	}

	close() {
		this.sidenav.close();
	}
}


src/app/app-routing.module.ts
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule, Routes } from '@angular/router';

import { HomeComponent } from './home/home.component';
import { ImageCropperComponent } from './image-cropper/image-cropper.component';
import { QuoteGeneratorComponent } from './quote-generator/quote-generator.component';
import { IpAddressFinderComponent } from './ip-address-finder/ip-address-finder.component';
import { MarkdownPreviewerComponent } from './markdown-previewer/markdown-previewer.component';

const routes: Routes = [
	{ path: 'ip-address-finder', component: IpAddressFinderComponent },
	{ path: 'quote-generator', component: QuoteGeneratorComponent },
	{ path: 'image-cropper', component: ImageCropperComponent },
	{ path: 'markdown-previewer', component: MarkdownPreviewerComponent },
	{ path: '', component: HomeComponent }
]

@NgModule({
  declarations: [],
  imports: [RouterModule.forRoot(routes)],
  exports : [RouterModule]
})
export class AppRoutingModule { }


src/app/ip-address-finder/ip-address-finder.component.ts
import IP from '../../assets/IP.json';
import { Component } from '@angular/core';

@Component({
  selector: 'app-ip-address-finder',
  templateUrl: './ip-address-finder.component.html',
  styleUrls: ['./ip-address-finder.component.css']
})
export class IpAddressFinderComponent {

	/*
	result : Object containing the final result in the format expected with
		key : IP Adsress
		value : count of the times IP appears in input			
	*/
	result = [];
	input = IP;
	columnsToDisplay = ['index','address', 'count'];
	constructor() {
		this.addressFinder(IP);
	}
	
	addressFinder(IP) {
		/*
		Splitting each element by the tab character
		Followed by splitting the 3rd element of the returned array by the new line character
		The first element in the returned array is the IP Address
		*/
		var addressObject = new Object();
		IP.forEach(el =>{		
			var ip = el.split('\t')[2].split('\n')[0];
			if(addressObject.hasOwnProperty(ip) ) {
				// If ip already added, increment the counter
				addressObject[ip]++;
			}
			else {
				addressObject[ip] = 1;
			}
		});

		for (var el in addressObject) {
		    this.result.push({'address':el, 'count':addressObject[el]});
		}

		this.result.sort(function(a, b) {
		    return a[1] - b[1];
		});
	}
}


src/app/image-cropper/image-cropper.component.ts
import { Component } from '@angular/core';
import { HttpClient } from '@angular/common/http';

@Component({
  selector: 'app-image-cropper',
  templateUrl: './image-cropper.component.html',
  styleUrls: ['./image-cropper.component.css']
})
export class ImageCropperComponent {

	fileUrl;
	crop1;
	crop2;
	crop3;
	crop4;
	constructor(private http: HttpClient) {}

	onSelectFile(event) {
		if (event.target['files'] && event.target['files'][0]) {
			var reader = new FileReader();
			reader.readAsDataURL(event.target['files'][0]); // read file as data url
			reader.onload = (event) => {
				this.fileUrl = event.target['result'];
				var imgObject = new Image();
				imgObject.src = this.fileUrl;
				imgObject.onload = (imageEvent) =>{
					this.bufferCanvas(imgObject);
				}
			}	
		}
	}

	// Kept on hold owing to API experiencing issues at the time of development
	viewOnImgur(imageBase64) {
		let base64ContentArray = imageBase64.split(",") 
		// base64 content cannot contain whitespaces but nevertheless skip if there are!
		let mimeType = base64ContentArray[0].match(/[^:\s*]\w+\/[\w-+\d.]+(?=[;| ])/)[0]

		var blob = new Blob([imageBase64], {type: mimeType});
		const headers = { 'Authorization': 'Client-ID 6272ea6b290ade8' }
		
		const formData = new FormData();
    	formData.append('image', blob);
    	formData.append('title', 'test');
    	formData.append('type', 'base64');
    	
    	formData.append('name', 'practice image');
    	this.http.post("https://api.imgur.com/3/image",formData,{ headers }).subscribe(response =>{
			console.log(response);
		},error =>{
			console.log("upload error ", error);
		});
	}

	bufferCanvas(image) {
		var bufferCanvas = document.createElement('canvas');
		var bufferContext = bufferCanvas.getContext('2d');
		bufferCanvas.width = image.width;
		bufferCanvas.height = image.height;
		bufferContext.drawImage(image, 0, 0);
		this.cr1(bufferCanvas);
		this.cr2(bufferCanvas);
		this.cr3(bufferCanvas);
		this.cr4(bufferCanvas);
	}

	cr1(bufferCanvas) {
		var tnCanvas = document.createElement('canvas');
 		var tnCanvasContext = tnCanvas.getContext('2d');
 		tnCanvas.width = 504; tnCanvas.height = 700;

 		// drawImage(image, sx, sy, sWidth, sHeight, dx, dy, dWidth, dHeight);
 		tnCanvasContext.drawImage(bufferCanvas, 0, 0, 755, 450 ,0,0,755,450);
 		this.crop1 = tnCanvas.toDataURL();	
	}

	cr2(bufferCanvas) {
		var tnCanvas = document.createElement('canvas');
 		var tnCanvasContext = tnCanvas.getContext('2d');
 		tnCanvas.width = 504; tnCanvas.height = 700;

 		// drawImage(image, sx, sy, sWidth, sHeight, dx, dy, dWidth, dHeight);
 		tnCanvasContext.drawImage(bufferCanvas, 0, 0, 365, 450 ,0,0,365,450);
 		this.crop2 = tnCanvas.toDataURL();	
	}

	cr3(bufferCanvas) {
		var tnCanvas = document.createElement('canvas');
 		var tnCanvasContext = tnCanvas.getContext('2d');
 		tnCanvas.width = 504; tnCanvas.height = 700;

 		// drawImage(image, sx, sy, sWidth, sHeight, dx, dy, dWidth, dHeight);
 		tnCanvasContext.drawImage(bufferCanvas, 0, 0, 365, 212 ,0,0,365,212);
 		this.crop3 = tnCanvas.toDataURL();	
	}

	cr4(bufferCanvas) {
		var tnCanvas = document.createElement('canvas');
 		var tnCanvasContext = tnCanvas.getContext('2d');
 		tnCanvas.width = 504; tnCanvas.height = 700;

 		// drawImage(image, sx, sy, sWidth, sHeight, dx, dy, dWidth, dHeight);
 		tnCanvasContext.drawImage(bufferCanvas, 0, 0, 380, 380 ,0,0,380,380);
 		this.crop4 = tnCanvas.toDataURL();	
	}
}
